### PR TITLE
Remove some unwraps from waldecoder

### PR DESCRIPTION
**Pull request by bojanserafimov**
https://github.com/neondatabase/neon/issues/474

I turned wal serialization/deserialization unwraps into errors. Panicking is still possible from all the `Buf` index arithmetic in waldecoder (`advance` can panic), but I'm leaving that for another PR.

Also I tried the [no_panic](https://docs.rs/no-panic/latest/no_panic/) crate, but it doesn't provide any diagnostics, and it doesn't specify what it promises (can there be false positives?) so maybe it's too early for solutions like that.

### CODE REVIEW

**Review by LizardWizzard — COMMENTED** [review]
Thanks for the patch!

I've left some comments. Mostly about usage of `.context` and avoiding `anyhow::Result` when possible

File: pageserver/src/layered_repository/delta_layer.rs [hunk 1]
```diff
@@ -359,7 +359,7 @@ impl Layer for DeltaLayer {
                                 write!(&mut desc, " img {} bytes", img.len()).unwrap();
                             }
                             Ok(Value::WalRecord(rec)) => {
-                                let wal_desc = walrecord::describe_wal_record(&rec);
+                                let wal_desc = walrecord::describe_wal_record(&rec).unwrap();
```
[the remark this answers is not in the document]
Comment (bojanserafimov) [reply pageserver/src/layered_repository/delta_layer.rs:362]: We're in a closure. I guess I can make the closure return result, but it's already unwrapping something else above, and it's only a debugging util function.

[the remark this answers is not in the document]
Comment (LizardWizzard) [reply pageserver/src/layered_repository/delta_layer.rs:362]: Hm, indeed, closures with results are a bit tricky, lets leave it as is then

File: pageserver/src/walrecord.rs [hunk 2]
```diff
@@ -511,7 +512,7 @@ pub fn decode_wal_record(record: Bytes) -> DecodedWALRecord {
     // 1. Parse XLogRecord struct
 
     // FIXME: assume little-endian here
-    let xlogrec = XLogRecord::from_bytes(&mut buf);
+    let xlogrec = XLogRecord::from_bytes(&mut buf)?;
```
[the remark this answers is not in the document]
Comment (bojanserafimov) [reply pageserver/src/walrecord.rs:518]: This is the only thing being deserialized in this function, it's easy to infer that without context.

[the remark this answers is not in the document]
Comment (LizardWizzard) [reply pageserver/src/walrecord.rs:518]: It will be just "deserialize error" then. From logs it would be quite hard to guess when the error was born

**Review by LizardWizzard — APPROVED** [review]
There is one unresolved comment, otherwise looks good to me.

Usage of WalDecodeError might be further improved by specifying more concrete cases in enum, with addition of `source` field when necessary, so err.source() will point to a next error in the chain. For example see DiskBtreeError [here](https://github.com/neondatabase/neon/blob/02e5083695d0ed17f7dbb1ca852f504fca42fdcc/pageserver/src/layered_repository/disk_btree.rs#L90) and docs for thiserror [here](https://docs.rs/thiserror/latest/thiserror/)

**Pull request merged**